Add dictionary extension

DrChai opened the pull request:
## Description

<!-- A summary of your change. If you add a new extension or command, explain what it does. -->

## Screencast

<!-- If you add a new extension or command, include a screencast (or screenshot for straightforward changes). A good screencast will make the review much faster - especially if your extension requires registration in other services.  -->

## Checklist

- [ ] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [ ] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [ ] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [ ] I checked that files in the `assets` folder are used by the extension itself
- [ ] I checked that assets used by the `README` are placed outside of the `metadata` folder

thomaslombart reviewed (commented):
Hey there, could you please add a description and a screencast? That will make reviewing the PR a bit easier 😊

Thanks!

DrChai commented:
@thomaslombart I just added a description and some screenshots. Apologies for not checking the actual PR comments and assuming everything was done after running `ray publish`.😥

thomaslombart reviewed (changes requested):
Thanks for this PR 💪 
I've reviewed it and have some feedback.

### Configuration Screen

There's a flickering on the initial load because I haven't configured anything yet. Also, if I configure something I need to press the ESC key to go back which doesn't feel intuitive. Is it possible to get rid of that? I can see this is mainly used for the default and fallback languages. Why not moving them to the preferences? That way the screen can be removed.

https://user-images.githubusercontent.com/16003285/231463483-5e6adbcd-b20e-4196-8079-e8456bc47d91.mp4

### Empty screen

The empty screen doesn't do much when launching the command. Note that you can use the `EmptyView` component to customize the title and the description to invite the user to do something.

<img width="884" alt="CleanShot 2023-04-12 at 14 48 48@2x" src="https://user-images.githubusercontent.com/16003285/231463993-19bb36d2-f6b1-42dc-ab08-8590288eff46.png">

### Switching to metadata view

The list switches to a metadata view after the first item which feels a bit weird.

Since the first item is used to show the actions already available in the action panel, why not removing it? That way you can keep the metadata view only.

https://user-images.githubusercontent.com/16003285/231464790-df8e625d-8341-42ee-9bb1-1b9ed323aa5f.mp4

### Cleaning code

There are a bunch of commented lines and `console.debug` statements in the code. Could you remove them to make the code cleaner for future potential contributors?

---

Please, request a new review when you are ready. Feel free to contact me here or in the [Slack community](https://www.raycast.com/community) if you have any questions.

DrChai commented:
@thomaslombart Thank you for taking the time and effort to review my PR. I will go through each feedback and submit a new commit this week.

DrChai commented on extensions/dictionary/package.json:
```diff
@@ -0,0 +1,81 @@
+      "arguments": [
+        {
+          "name": "arguments-title",
+          "placeholder": "-set",
```
tried to play with the argument and see how it works.😅 will remove it in the next commit

DrChai commented on extensions/dictionary/src/engines/urban.ts:
```diff
@@ -0,0 +1,92 @@
+#### Example
+> ${encode(example)}
+
+#### 👍   ${def.thumbs_up}   |   👎   ${def.thumbs_down}
```
Currently, when the input query is a bit longer, accessory values would also get truncated.
<img width="843" alt="Screenshot 2023-04-12 at 5 12 08 PM" src="https://user-images.githubusercontent.com/1713574/231586344-091947bd-e8b8-46d1-8199-1b1e22f315ea.png">

DrChai commented:
So I fixed the configuration view that appeared on the initial load. The issue was actually caused by an unset engine variable on the dropdown menu, it now uses Google as the engine and 'en' as the language.
 I have also considered using a metadata view, but it appears that many definitions not come with related examples:
 
<img width="833" alt="Screenshot 2023-04-13 at 2 17 31 AM" src="https://user-images.githubusercontent.com/1713574/231698406-e64cd1be-b735-43e3-9607-614d04bf6749.png">
There would be too much white space on the metadata view.😔

I switched to using `useCachedState`. I tried to create a custom hook that also subscribes to the cache on view mount, triggering a rerender when specific keys update.   Turns out the `useCachedState` already includes this function 👍, although it is not mentioned in the documentation. 
lastly, I added some custom messages on the emptyview:
<img width="815" alt="Screenshot 2023-04-13 at 3 35 42 AM" src="https://user-images.githubusercontent.com/1713574/231702819-e2c31f32-8205-40fe-9021-c3cff0f80d56.png">

thomaslombart commented:
Thanks for addressing the issues 💪  
I've reviewed it again and have spotted other things.

### Wrong title for empty screen

The title says `The word(s) will be translated to 'en'` when using `Urban Dictionary`. As far as it looks, there is no translation with the urban dictionary, right?

<img width="891" alt="CleanShot 2023-04-19 at 10 44 43@2x" src="https://user-images.githubusercontent.com/16003285/233021140-6e34f888-1add-4863-8f5a-a256cff6bada.png">

Also for Google Translation, ideally, the world should correspond to the plain language. For example: `The word(s) will be translated to 'en'` would become `The word(s) will be translated to English`.

### Empty screen while searching

The empty screen still shows up when you start searching for a word. If a user starts searching the screen should disappear.

https://user-images.githubusercontent.com/16003285/233020992-d28c4a88-c05a-45a0-8803-b887bc6ee771.mp4

### Urban dictionary result

I found the urban dictionary detail view to be a bit weird as the name gets repeated many times. Why is that and is it possible  to make it less redundant?

<img width="874" alt="CleanShot 2023-04-19 at 10 52 17@2x" src="https://user-images.githubusercontent.com/16003285/233022447-ee72bad3-36a2-44c3-9f08-6ace94a128da.png">

DrChai commented:
@thomaslombart 

> The empty screen still shows up when you start searching for a word. If a user starts searching the screen should disappear.

It appears that the default empty view will always render as long as the list is empty, so I tried to prevent my custom empty view from rendering when the query is empty. However, the default view is still being rendered.
```typescript
<List>
...

  {
        query?null:<List.EmptyView title={emptyTitle} description={emptyDescription} />
  }
      
 </List>
```
https://user-images.githubusercontent.com/1713574/233427128-12d6e170-5a30-4b5b-b8bd-6e1e70dd2b58.mov

Although it is not displayed when `isLoading`, here I applied a debounce on input so fetch only starts when the user stop typing. Is there another way to control the empty view?

> I found the urban dictionary detail view to be a bit weird as the name gets repeated many times. Why is that and is it possible to make it less redundant?

Urban Dictionary always shows multiple definitions that users submitted for each term. So basically it has two endpoints, one for searching and another for browsing the specific term. Here is how it's actually implemented on its website:
<img width="842" alt="Screenshot 2023-04-20 at 12 42 06 PM" src="https://user-images.githubusercontent.com/1713574/233432745-0eb48f63-2207-429c-a76c-9dd8b2e0a257.png">
<img width="660" alt="Screenshot 2023-04-20 at 12 41 32 PM" src="https://user-images.githubusercontent.com/1713574/233432871-cc69043d-2ba5-4b14-af05-8a1da81bda7f.png">

>The title says The word(s) will be translated to 'en' when using Urban Dictionary. As far as it looks, there is no translation with the urban dictionary, right?

Yes, this message is a generic template and is shown on whatever the dictionary selected, I will submit a new commit to allow some dictionaries like Urban to have their own messages.

thomaslombart commented:
> Although it is not displayed when isLoading, here I applied a debounce on input so fetch only starts when the user stop typing. Is there another way to control the empty view?

Using `isLoading` is the way to go to for that. The no results screen won't show while `isLoading` is `true`.

DrChai commented:
> > Although it is not displayed when isLoading, here I applied a debounce on input so fetch only starts when the user stop typing. Is there another way to control the empty view?
> 
> Using `isLoading` is the way to go to for that. The no results screen won't show while `isLoading` is `true`.

Hi @thomaslombart, I think I got an unexpected result. As you can see from the screenshot, even when `isLoading` is updated to `true` and the loading bar pops up, the empty view is still present.
<img width="1161" alt="Screenshot 2023-04-21 at 8 10 57 AM" src="https://user-images.githubusercontent.com/1713574/233634521-bdf1bcb2-756e-4ff0-b1ff-242a319b59b6.png">
For me, the empty view only dismisses when the isLoading variable is set to true initially.

thomaslombart commented:
That's happening because the component got rendered initially with `isLoading` set to `false.. To prevent that, you need to make sure you on the first render that `isLoading` is set to `true`.

thomaslombart reviewed (approved):
Looks good to me! As said in direct messages, there are certain patterns that could be improved but let's see if users actually have feedback about them 😊
